src/main/kotlin/org/rust/ide/docs/RsDocumentationProvider.kt

```diff
@@ -17,89 +17,101 @@ class RsDocumentationProvider : AbstractDocumentationProvider() {
-            is RsEnumItem -> listOf(whereClause)
-            is RsEnumVariant -> listOf(tupleFields)
-            is RsTraitItem -> listOf(whereClause)
-            is RsTypeAlias -> listOf(typeReference, typeParamBounds, whereClause)
-            is RsConstant -> listOf(expr, typeReference)
-            is RsSelfParameter -> listOf(typeReference)
-            else -> listOf(navigationElement)
-        }
+        // Remove trailing spaces, comments and attributes
+        val signatureStartElement = generateSequence(decl.firstChild) { it.nextSibling }
+            .dropWhile { it is PsiWhiteSpace || it is PsiComment || it is RsOuterAttr }
+            .firstOrNull()
+
+        val signatureStart = signatureStartElement?.startOffsetInParent ?: return id.text
+        val idStart = id.ancestors.takeWhile { it != decl }.sumBy { it.startOffsetInParent }
 
         // pick (in order) elements we should stop at
         // if they all fail, drop down to the end of the id element
-        val idElement = if (this is RsSelfParameter) self else navigationElement
         val signatureEnd = stopAt
-                .filterNotNull().firstOrNull()
-                ?.let { it.startOffsetInParent + it.textLength }
-                ?: idElement.startOffsetInParent + idElement.textLength
+            .filterNotNull().firstOrNull()
+            ?.let { it.startOffsetInParent + it.textLength }
+            ?: idStart + id.textLength
 
-        val identStart = idElement.startOffsetInParent
-        val identEnd = identStart + idElement.textLength
-        check(signatureStart <= identStart && identStart <= identEnd &&
-            identEnd <= signatureEnd && signatureEnd <= textLength)
+        val idEnd = idStart + id.textLength
+        check(signatureStart <= idStart && idStart <= idEnd &&
+            idEnd <= signatureEnd && signatureEnd <= decl.textLength)
 
-        val beforeIdent = text.substring(signatureStart, identStart).escaped
-        val afterIdent = text.substring(identEnd, signatureEnd)
+        val beforeIdent = decl.text.substring(signatureStart, idStart).escaped
+        val afterIdent = decl.text.substring(idEnd, signatureEnd)
             .replace("""\s+""".toRegex(), " ")
             .replace("( ", "(")
             .replace(" )", ")")
             .replace(" ,", ",")
             .trimEnd()
             .escaped
 
-        return "$beforeIdent<b>$name</b>$afterIdent"
+        return "$beforeIdent<b>${id.text}</b>$afterIdent"
     }
 
-    private fun RsNamedElement.getQuickNavigateInfo(): String =
-        "${formatSignature()}$locationString"
+    private val RsFunction.signatureText: String?
+        get() = createSignatureText(this, identifier, listOf(whereClause, retType, valueParameterList))
 
+    private val RsPatBinding.owner: PsiElement?
+        get() = ancestors
+            .drop(1).dropWhile { it is RsPat || it is RsMatchPat || it is RsPatField }
```

Comment:
Yes, that's intentional. Although the name may be misleading. `RsMatchPat` is not a pattern because you can't stick it anywhere they pattern can go:

```
let (None | Some(x)) = None; // syntax error
let (None if cond) = None; // ditto
```

The same goes for `RsPatField`.